_posts/pl/2021-09-31-zarzadzanie-stanem-aplikacji-frontendowej.md

```diff
@@ -0,0 +1,225 @@
+
+Załóżmy, że chcemy stworzyć prosty program pozwalający na zapisywanie sposobów na pokonanie nudy. Jednym z możliwych do podjęcia działań będzie wywołanie API, które na każde żądanie odpowie pomysłem na jakąś aktywność. Drugą opcją będzie dodanie własnej koncepcji poprzez prosty formularz. Zacznijmy zatem.
+
+### Store
+Na samym początku procesu, warto uzmysłowić sobie, co tak naprawdę chcemy przechowywać w naszym store, a potem powiedzieć to Typescriptowi tworząc interfejs. W naszym przypadku będziemy przechowywali w nim jedynie listę czynności potencjalnie zabijających nudę, jednak nawet w tak prostym przypadku warto zwrócić uwagę na to, jaką strukturę planujemy nadać store'owi. Początkowo najprościej jest operować na płaskiej strukturze, w której wszystkie dane znajdują się na tym samym poziomie.
+Niestety przy rozwijaniu aplikacji takie podejście staje się bardzo nieefektywne, zarówno pod względem organizacji, jak i, z czasem, również wydajności. Zdecydowanie lepiej jest podzielić store na tak zwane "slice'y", czyli wycinki danych, na przykład dzieląc store według funkcjonalności aplikacji. Zobaczmy takie podejście na przykładzie:
+
+```typescript
+export interface State {
+    activitiesState: ActivityItemsState;
+}
+
+export interface ActivityItemsState {
+    activities: ActivityItemModel[];
+}
+
+export interface ActivityItemModel {
+    name: string;
+    participants: number;
+}
+
+```
+
+Przejdźmy po powyższych interfejsach po kolei. Pierwszy z nich mówi, że spodziewamy się mieć jeden wycinek danych (slice) typu `ActivityItemsState`, w którym będą przechowywane wszystkie informacje dotyczące aktywności dodanych do aplikacji. Następnie definiujemy, co tak naprawdę znajduje się w tym wycinku danych - jest to tablica obiektów typu `ActivityItemModel`, czyli informacji o różnych aktywnościach. Ostatni interfejs to już wyłącznie definicja modelu biznesowego - w naszej aplikacji będziemy mieli do czynienia z nazwą czynności oraz możliwą liczbą jej uczestników.
+
+Taki podział store zapewnia bardzo łatwą jego rozszerzalność, gdyż w przypadku dodania nowej funkcjonalności wystarczy dopisać nowy wycinek danych do store. W bardzo dużych aplikacjach, gdzie obiekt stanu aplikacji jest już dość potężny, to podejście ma jeszcze jedną zaletę - pozwala wykorzystać _lazy loading_, to znaczy wczytywać tylko konkretne wycinki store, potrzebne do aktualnie przetwarzanej części aplikacji. 
+
+### Akcje
+Potrzebujemy teraz możliwości wywoływania akcji, żeby powodować zmiany w store. Aplikacja będzie przewidywała dwa zdarzenia: wystosowanie żądania do API po pomysł na jakieś zajęcie oraz dodanie aktywności do listy. Wygląda to następująco:
```

Comment:
czepiam się, wiem. akcje nie służą do 'wywoływania zmian w store'. akcje służą do zgłaszania przeszłych faktów. to czy to wpłynie na stan store czy nie. w jaki sposób i w ilu miejscach to trochę kwestia logiki w reducerach. gdyby to było tak, że każda operacja jest definiowana jako procedura to komponenty by zgłaszały konkretne 'zrób coś' i w praktyce jaka wartość ze store skoro zawsze trzeba je zmieniać razem z logiką?

czy nie ciekawsze jest np. 'Użyszkodnik zaproponował aktywność' i to od reducera zależy czy ją doda/jak doda/co zrobi? wiem, że to trywialny przykład. może to jest właśnie problem?

Reply:
zobacz sobie np. na  "The only way to change the state is to emit an action, an object describing what happened." za https://redux.js.org/understanding/thinking-in-redux/three-principles

Akcja opisuje jedynie co się wydarzyło. A Twoje akcje mówią co zrobić "Dodaj" i "Pobierz"